<img src="../.images/ChatbotSentiment/bannerUgentDwengo.png" alt="BannerUGentDwengo" style="width:250px;"/>

<div style='color: #690027;' markdown="1">
    <h1>SENTIMENTANALYSE</h1> 
    <h1>deel 1</h1> 
</div>

<div class="alert alert-box alert-success">
Taaltechnologen doen een beroep op machine learning modellen om bij gegeven teksten onderzoek te doen naar sentimentwoorden. In deze notebook maak je kennis met de principes van hun onderzoek. <br> Het gebruik van technologie wordt steeds toegankelijker voor werknemers in een niet-technologische sector, zoals taalkundigen, communicatiewetenschappers, historici en juristen. <br>Dankzij de zeer toegankelijke programmeertaal Python zal ook jij enkele mogelijkheden van technologie ontdekken.
</div>

<div class="alert alert-block alert-info"> 
Python is vaak zeer intuïtief in gebruik en bovendien zo populair dat er heel wat modules voorhanden zijn die men vrij kan gebruiken. In een module zitten heel wat functies vervat die ervaren informatici reeds voor jou hebben geprogrammeerd. 
</div>

<div class="alert alert-block alert-warning"> 
Een notebook is een interactief document dat ook programmacode bevat. In deze notebook zie je verschillende grijze vakken staan. Dat zijn code-cellen die Pythoncode bevatten. Zo'n code-cel wordt voorafgegaan door <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">In [ ]</span>. Bij het uitvoeren van de code verschijnt er een * tussen de  <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">[ ]</span>. Eens de code is uitgevoerd, staat er een getal tussen de  <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">[ ]</span>.<br>
<b>Om de code in een code-cel uit te voeren, klik je op de cel en voer je die uit met de knop 'Run' bovenaan in het menu. <img src="../.images/ChatbotSentiment/runknop.jpg" alt="Banner" align="right" style="width:70px;"/>
</div>

Heb je het begrepen? Voer dan de volgende code-cel uit door in het menu bovenaan op 'Run' te klikken. 

In [1]:
print("Ik snap het!")

Ik snap het!


<div style='color: #690027;' markdown="1">
    <h2>1. Werken met strings en lijsten</h2> 
</div>

### Een string

De **uitvoer** die verscheen, was: Ik snap het! Deze uitvoer bestaat uit letters, twee spaties en een leesteken. 

<div class="alert alert-block alert-info"> 
<span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">Ik snap het!</span> is een <b>string</b>: een opeenvolging van karakters, zoals letters, leestekens en spaties. Je kan een string ook herkennen aan <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">" "</span>.<br> 
<span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">print()</span> is een <b>functie</b> waarmee je aangeeft dat er iets moet worden getoond op het scherm.
</div>

### Een index

Neem nu een andere string: "Hello, world!" Elk karakter in een string heeft een bepaalde plaats. In Python komt de plaats van een karakter in een string overeen met een **index**. Men begint daarbij te tellen vanaf 0.<br>

Bijvoorbeeld:
-  De letter "H" is het eerste karakter. De index van "H" is 0.
-  De letter "e" in de string staat op de tweede plaats. De index van "e" is 1.
-  De eerste "o" vind je op de vijfde plaats. De index van de eerste "o" is 4.
-  De spatie is het zevende karakter van deze string. De index van de spatie is 6. 

<div>
<img src="../.images/ChatbotSentiment/helloworldIndex.png" alt="Banner" align="center" style="width:500px;"/>
</div>

#### Oefening
-  Welk karakter heeft 7 als index? 
-  Welke letter staat op de twaalfde plaats? 

Je vraagt de antwoorden van de oefening op met Python door de volgende code-cel uit te voeren.

In [2]:
zin = "Hello, world!"
print(zin[7])
print(zin[11])

w
d


<div class="alert alert-block alert-info"> 
<span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">zin</span> is een <b>variabele</b> die verwijst naar de string <span style="background-color:whitesmoke; font-family:consolas; font-size:1em;">"Hello, world!"</span>. De naam die je aan de variabele geeft, kies je zelf.
</div>

Beschouw nog een andere string: "Programmeren is voor iedereen." 

#### Oefening
-  Wat zijn de indices van de letters "d" en "P"?
-  Wat is de index van de tweede spatie?
-  Welk karakter heeft 14 als index? 

Controleer dit laatste eens met Python. Geef een geschikte naam aan de variabele.

In [3]:
nieuwezin = "Programmeren is voor iedereen"
print(nieuwezin[14])

s


<div class="alert alert-block alert-info"> 
Elk karakter in een string heeft een <b>index</b> die bepaalt het hoeveelste karakter het is in die string. 
</div>

### Strings samenvoegen

Door gebruik te maken van de +-operator kun je strings samenvoegen. Test dit uit door de volgende code-cellen uit te voeren. 

In [4]:
# strings concateneren
woord1 = "sentiment"
woord2 = "analyse"
woord = woord1 + woord2
print(woord)

sentimentanalyse


<div class="alert alert-block alert-info"> 
Tekst die na een <b>#</b> staat, is commentaar om de code beter te begrijpen. Commentaar wordt genegeerd door de computer.
</div>

In [5]:
# strings concateneren
deel1 = "Onderz"
deel2 = "oek"
deel3 = " "
deel4 = "d"
deel5 = "oen"
deel6 = "?"
geheel = deel1 + deel2 + deel3 + deel4 + deel5 + deel6
print(geheel)

Onderzoek doen?


Voeg zelf eens enkele strings samen met Python:

<div class="alert alert-block alert-info">
Bij het samenvoegen of <b>concateneren</b> van strings met de +-operator, ontstaat een nieuwe string.
</div>

### Lijsten

Voer de volgende code-cel uit.

In [6]:
lijst = ["boter", "kan", "boos", "zielepoot"]    # lijst met vier elementen

Kun je raden wat het element in deze lijst is met index 0? En met index 3?

In [7]:
print(lijst[0])
print(lijst[3])

boter
zielepoot


Je begrijpt nu wat de volgende begrippen inhouden:
- een string;
- een index van een karakter in een string;
- het samenvoegen van strings tot één nieuwe string (concateneren);
- een index van een element van een lijst.

<div style='color: #690027;' markdown="1">
    <h3>Logische operatoren</h3> 
</div>

<table>
 <thead>
    <tr>
      <th>&nbsp;</th>  
      <th><p align="center">Wiskunde</th>
      <th>&nbsp;</th>
      <th><p align="center">Python</th>    
  </thead>
   <tr> <td> <p align="left">groter dan          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; >   <td> <td> 
      <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; >   
  <tr> <td> <p align="left">kleiner dan           <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; <   <td> <td> 
      <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; <    
  <tr> <td>  <p align="left">gelijk aan  <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; =   <td> <td> 
      <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; ==  
  <tr> <td> <p align="left">verschillend van          <td> <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; &#8800;  <td> <td> 
      <p align="center">&nbsp;&nbsp;&nbsp;&nbsp; !=    
      
</table>

In [8]:
lijst[2] == "boos"

True

In [9]:
lijst[1] == "kat"

False

In [10]:
lijst[1] != "kat"

True

<div style='color: #690027;' markdown="1">
    <h2>2. Sentimentanalyse</h2> 
</div>

### Principes van sentimentanalyse

Voor **regelgebaseerde** sentimentanalyse maak je gebruik van een (bestaand) **lexicon** met daarin woorden gekoppeld aan hun **polariteit** (positief, negatief of neutraal), dus een woordenboek van sentimentwoorden. 

'Blij' bijvoorbeeld heeft een positieve polariteit, 'thuisbankieren' een neutrale en 'boos' een negatieve polariteit. In het lexicon wordt de polariteit weergegeven door een getal tussen -2 en 2. 

<div>
<img src="../.images/ChatbotSentiment/schaal.png" alt="Banner" align="center" style="width:500px;"/>
</div>

De polariteit van een tekst wordt gegeven door de som van de polariteiten van de sentimentwoorden in die tekst.

Voor je sentimentwoorden uit een lexicon kunt matchen met de gegeven tekst (de data) moet je de data 
-  1) *inlezen*; 
-  2) *preprocessen*, d.w.z. voorbereiden voor *lexicon matching* of voor *machine learning*. 

Preprocessing omvat alle stappen die nodig zijn om de data voor te bereiden op wat volgt, of dat nu een eenvoudige lexicon matching is, dan wel een ingewikkeld machine learning systeem dat op de data getraind zal worden. <br>Hieronder vind je een oplijsting van veelvoorkomende preprocessing stappen.


#### Preprocessing

* **Lowercasing:** alle hoofdlettertekens worden vervangen door kleine letters.
* **Tokenisering:** alle zinnen worden in betekenisvolle eenheden of 'tokens' gesplitst, zoals woorden en leestekens.
* **Part-of-speech tagging:** aan elk token wordt de grammaticale woordcategorie toegekend, zoals adjectief of symbool.
* **Lemmatisering:** alle tokens worden omgezet naar hun lemma of woordenboekvorm (bv. een substantief komt in een woordenboek in het enkelvoud voor en van een werkwoord vind je er de infinitief). Die woordenboekvorm wordt dan opgezocht in het lexicon. 

#### Voorbeeld:
Gegeven zin:    
-  De spelletjes waren toffe ijsbrekers.

Lowercasing:    
-  de spelletjes waren toffe ijsbrekers.

Tokens:         
-  'de' 'spelletjes' 'waren' 'toffe' 'ijsbrekers' '.' 

Part-of-speech: 
-  'de': lidwoord;
-  'spelletjes': substantief; 
-  'waren': werkwoord;
-  'toffe': adjectief;
-  'ijsbrekers': substantief;
-  '.': leesteken (symbool).

Lemma's: 'de', 'spel', 'zijn', 'tof', 'ijsbreker', '.'

Polariteit:
-  De polariteiten van de lemma's worden opgezocht in het lexicon, lidwoorden en leestekens zijn daarbij niet van belang.  
-  'spel' heeft polariteit 1, 'zijn' heeft polariteit 0, 'tof' heeft polariteit 0,8 en 'ijsbreker' 0.
-  De polariteit van de gegeven zin is de som van deze polariteiten, dus 1,8. De zin roept een positief sentiment op. 

### Modules importeren 

Nu je dit weet, kun je bijna aan de slag. Je laadt eerst twee Python-modules en een lexicon in. <br>
Voer daartoe de code-cel hieronder uit. De code in deze cellen hoef je niet in detail te begrijpen.

In [11]:
# module importeren
import pickle      # voor lexicon
from colorama import Fore, Back   # om in kleur te kunnen printen

In [12]:
# lexicon inlezen 
with open("new_lexicondict.pickle", "rb") as f: # bestand lexicondict.pickle in map data bevat het sentimentlexicon
    lexicondict = pickle.load(f)

# lexicon tonen
print(lexicondict)

{'retorisch': {'postag': ['ADJ'], 'polarity': [0.0]}, 'gezwind': {'postag': ['ADJ'], 'polarity': [0.6]}, 'evenwichtig': {'postag': ['ADJ'], 'polarity': [1.25]}, 'modaal': {'postag': ['ADJ'], 'polarity': [0.4]}, 'digitaal': {'postag': ['ADJ'], 'polarity': [0.0]}, 'fout': {'postag': ['ADJ', 'NOUN'], 'polarity': [-0.5, -2.0]}, 'onverdeeld': {'postag': ['ADJ'], 'polarity': [0.1]}, 'wulps': {'postag': ['ADJ'], 'polarity': [0.7]}, 'bemoeiziek': {'postag': ['ADJ'], 'polarity': [-1.35]}, 'grootscheeps': {'postag': ['ADJ'], 'polarity': [0.2]}, 'vorig': {'postag': ['ADJ'], 'polarity': [0.0]}, 'creatief': {'postag': ['ADJ'], 'polarity': [1.2]}, 'tegenoverliggend': {'postag': ['ADJ'], 'polarity': [-0.2]}, 'schalks': {'postag': ['ADJ'], 'polarity': [0.0]}, 'kurkdroog': {'postag': ['ADJ'], 'polarity': [0.0]}, 'kwistig': {'postag': ['ADJ'], 'polarity': [-0.1]}, 'verknipt': {'postag': ['ADJ'], 'polarity': [-1.2]}, 'tiranniek': {'postag': ['ADJ'], 'polarity': [-0.3]}, 'mondig': {'postag': ['ADJ'], 'pol

Een stukje van het lexicon:<br>  'retorisch': {'postag': ['ADJ'], 'polarity': [0.0]}, <br> 'gezwind': {'postag': ['ADJ'], 'polarity': [0.6]}, <br>'evenwichtig': {'postag': ['ADJ'], 'polarity': [1.25]},<br> 'modaal': {'postag': ['ADJ'], 'polarity': [0.4]},<br> 'digitaal': {'postag': ['ADJ'], 'polarity': [0.0]}, <br>'fout': {'postag': ['ADJ', 'NOUN'], 'polarity': [-0.5, -2.0]}, ... , <br>'analfabeet': {'postag': ['NOUN'], 'polarity': [-1.0]}, ..., <br>'stigmatiseren': {'postag': ['VERB'], 'polarity': [-2.0]},  ...

Het lexicon geeft de part-of-speech tag ('postag') en de polariteit ('polarity') van de woorden in het lexicon.

Uit het lexicon leid je af:
-  het woord 'retorisch' is een adjectief dat qua sentiment een neutrale polariteit heeft;
-  het woord 'gezwind' is een adjectief dat qua sentiment een positieve polariteit heeft;
-  het woord 'evenwichtig' is een adjectief dat qua sentiment ook een positieve polariteit heeft, maar het wordt positiever aangevoeld dan 'gezwind';
-  het woord 'fout' kan zowel een adjectief als een substantief zijn, beide met een negatieve polariteit, maar als substantief wordt het als meer negatief aangevoeld dan als adjectief.
- het woord 'analfabeet' is een substantief dat qua sentiment een negatieve polariteit heeft;
- het woord 'stigmatiseren' is een werkwoord dat qua sentiment een negatieve polariteit heeft, meer negatief dan 'analfabeet';

Noteren in de code:
-  `lexicondict["retorisch"]["postag"]` is `['ADJ']`
-  `lexicondict["retorisch"]["polarity"]` is `[0.0]`
-  `lexicondict["fout]["postag"]` is `['ADJ', 'NOUN']`
-  `lexicondict["fout"]["polarity"]` is `[-0.5, -2.0]`

Je zou het lexicon dus ook kunnen weergeven als:<br>
retorisch ADJ 0.0<br>
gezwind ADJ 0.6<br>
evenwichtig ADJ 1.25<br>
modaal ADJ 0.3<br>
digitaal ADJ 0.0<br>
fout ADJ -0.5<br>
fout NOUN -2.0

#### Oefening:

Zoek in het lexicon op:
-  de "postag" van 'kwekkebekken';
-  de "postag' van  'aanraden';
-  de "polarity" van 'jolijt';
-  de "polarity" van 'konkelen'.

Tip: zoek met CTRL-F

**Merk op** dat de woordsoorten in het Engels teruggegeven worden. 

-  *'NOUN'* staat voor een zelfstandig naamwoord of substantief; 
-  *'ADJ'* voor een bijvoeglijk naamwoord of adjectief; 
-  *'ADV'* voor een bijwoord; 
-  *'DET'* voor een lidwoord; 
-  *'VERB'* voor een werkwoord; 
-  *'AUX' voor een hulpwerkwoord;
-  *'PRON'* voor een voornaamwoord;
-  *'PROPN'* voor een eigennaam, enz. 
-  Aan leestekens wordt *'SYM'* toegekend.

De woorden die in het lexicon staan, noemt men *sleutels* of *keys* in Python. Je kan ze opvragen via de volgende code-cel. 

In [13]:
lexicondict.keys()

dict_keys(['retorisch', 'gezwind', 'evenwichtig', 'modaal', 'digitaal', 'fout', 'onverdeeld', 'wulps', 'bemoeiziek', 'grootscheeps', 'vorig', 'creatief', 'tegenoverliggend', 'schalks', 'kurkdroog', 'kwistig', 'verknipt', 'tiranniek', 'mondig', 'deels', 'verrassend', 'tomeloos', 'schadelijk', 'luguber', 'protestants', 'uiterst', 'eerloos', 'loodgrijs', 'zogeheten', 'harmonieus', 'fijnbesnaard', 'plechtig', 'gemaakt', 'ondubbelzinnig', 'overijverig', 'psychisch', 'vast', 'schoon', 'oubollig', 'basaal', 'filantropisch', 'steengoed', 'kouwelijk', 'sefardisch', 'tussenliggend', 'zelfvoldaan', 'geheimzinnig', 'impulsief', 'definitief', 'aandoenlijk', 'storend', 'plechtstatig', 'koortsachtig', 'doortimmerd', 'humeurig', 'doodgeboren', 'bekakt', 'categoriaal', 'consistent', 'fit', 'spoedig', 'direct', 'overspannen', 'zuurverdiend', 'ongelezen', 'onbewogen', 'fortuinlijk', 'verkeken', 'pikant', 'onhebbelijk', 'overdadig', 'heimelijk', 'krengerig', 'milieubewust', 'dakloos', 'consequent', 'blut'

Zo, je bent klaar voor een toepassing. Stap 1: de data (gegevens) inlezen en bekijken. 

<div style='color: #690027;' markdown="1">
    <h2>3. Toepassing</h2> 
</div>

Voor deze opdracht zal je werken met een **klantenreview**. 

### De data 

Voer de volgende code-cel uit om de review in te lezen en vervolgens te bekijken.

In [14]:
review = "Nieuw concept in Gent, maar dat kan volgens mij toch beter. De meeste cornflakes waren gewoon de basic soorten. Ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. En als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie."

In [15]:
print(review)

Nieuw concept in Gent, maar dat kan volgens mij toch beter. De meeste cornflakes waren gewoon de basic soorten. Ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. En als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie.


Je bent klaar voor stap 2. In wat volgt voer je de preprocessing uit op de review.

<div style='color: #690027;' markdown="1">
    <h2>4. Preprocessing</h2> 
</div>

### Lowercasing
In deze stap zet je de tekst in de review om naar kleine letters. Lowercasing is nodig omdat woorden zonder hoofdletter in het lexicon staan.

De variabele `review_kleineletters` verwijst naar deze omgezette tekst.

In [16]:
# zet tekst van de review om naar tekst in kleine letters
review_kleineletters = review.lower()  # review met kleine letters schrijven

In [17]:
# toon resultaat van lowercasing
print(review_kleineletters)

nieuw concept in gent, maar dat kan volgens mij toch beter. de meeste cornflakes waren gewoon de basic soorten. ook wat duur voor de hoeveelheid die je krijgt, vooral met de toppings zijn ze zuinig. en als je ontbijt aanbiedt, geef de mensen dan toch ook wat meer keuze voor hun koffie.


### Tokenisatie
Nu zal je de review in woorden en leestekens opsplitsen m.b.v. de computer, dat gebeurt op basis van spaties.
Deze woorden en leestekens zijn **tokens**.

Om de tokens automatisch te laten genereren, is het nodig dat de woorden en leestekens van elkaar gescheiden zijn door een spatie. Na elke spatie kan dan een nieuwe token worden gegenereerd.
Bv. Hello, world!   wordt eerst geschreven als Hello , world ! en de tokens zijn dan: 'Hello', ',', 'world' en '!'.

Je zal de review dus eerst wat moeten aanpassen: voor en na elk leesteken moet zeker een spatie aanwezig zijn.  

#### Spaties invoeren vóór elk leesteken
In deze stap plaats je in de review een spatie voor (en na) elk leesteken.

De variabele `review_spatie` verwijst naar deze omgezette tekst.

In [18]:
leestekens = ['!', '"', '#', '$', '%', '&', "'", '(', ')', '*', '+', ',', '-', '.', '/', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', '{', '|', '}', '~']

In [19]:
review_spatie = " "     # lege string
for karakter in review_kleineletters:
    if karakter not in leestekens:
        review_spatie = review_spatie + karakter
    else:
        review_spatie = review_spatie + " " + karakter + " "     # spatie voor en na een leesteken

In [20]:
# toon resultaat van spaties toevoegen
print(review_spatie)

 nieuw concept in gent ,  maar dat kan volgens mij toch beter .  de meeste cornflakes waren gewoon de basic soorten .  ook wat duur voor de hoeveelheid die je krijgt ,  vooral met de toppings zijn ze zuinig .  en als je ontbijt aanbiedt ,  geef de mensen dan toch ook wat meer keuze voor hun koffie . 


In [21]:
# zet tekst van de review om naar tekst in kleine letters
tokens = review_spatie.split()  # review met kleine letters schrijven

In [22]:
# toon resultaat van lowercasing
print(tokens)

['nieuw', 'concept', 'in', 'gent', ',', 'maar', 'dat', 'kan', 'volgens', 'mij', 'toch', 'beter', '.', 'de', 'meeste', 'cornflakes', 'waren', 'gewoon', 'de', 'basic', 'soorten', '.', 'ook', 'wat', 'duur', 'voor', 'de', 'hoeveelheid', 'die', 'je', 'krijgt', ',', 'vooral', 'met', 'de', 'toppings', 'zijn', 'ze', 'zuinig', '.', 'en', 'als', 'je', 'ontbijt', 'aanbiedt', ',', 'geef', 'de', 'mensen', 'dan', 'toch', 'ook', 'wat', 'meer', 'keuze', 'voor', 'hun', 'koffie', '.']


Je krijgt dus zo een lijst van de tokens.

Als de tekst getokeniseerd is, kan je aan elk token een **part-of-speech tag** toekennen, bijvoorbeeld voor een woord is dat de woordsoort.<br>Bij het opzoeken van een token in het lexicon is het belangrijk dat je daarbij ook de part-of-speech tag controleert. Sommige woorden kunnen immers zowel als substantief of als adjectief gebruikt worden.

Een token opzoeken in het lexicon doe je in zijn woordenboekgedaante. Elk token moet dus gelemmatiseerd worden, m.a.w. teruggebracht worden naar zijn **lemma** of woordenboekvorm, zoals enkelvoud voor een substantief en de infinitief voor een werkwoord. Het opzoeken van zo'n woordenboekvorm in het lexicon kan dan **geautomatiseerd** worden.

### Part-of-speech tagging en lemmatisering

Je wil dus een lijst van de part-of-speechs tags en van de lemma's die bij de tokens horen, en deze lemma's kan je dan opzoeken in het lexicon.

Het eerste token is 'nieuw'. De basisvorm of lemma is `nieuw` en dit is een adjectief, ADJ.

Neem het token ','. De basisvorm of lemma is `,` en dit is een leesteken, dus een symbool, SYM.

Neem het token 'kan'. De basisvorm of lemma is `kunnen` en dit is een werkwoord, VERB.

Neem het token 'soorten'. De basisvorm of lemma is `soort` en dit is een substantief, NOUN.

**Merk op** dat het lemma van tokens zoals *'waren'*, *'soorten'*, *'krijgt'*, *'toppings'*, *'aanbiedt'*, *'geef'* en *'mensen'* niet hetzelfde is als het token. Bij andere tokens is dat wel het geval.

Maak lijsten van de lemma's en de part-of-speech tags. De volgorde in deze lijsten is van belang.<br>
Tip: vertrek van de lijst van de tokens en gebruik 'copy-paste' (CTRL-C, CTRL-V), pas dan enkel de lemma's aan die verschillen van het token. 

In [23]:
postags = ['ADJ', 'NOUN', 'ADP', 'PROPN', 'SYM', 'CCONJ', 'PRON', 'VERB', 'ADP', 'PRON', 'ADV', 'ADJ', 'SYM', 'DET', 'ADV', 'NOUN', 'AUX', 'ADJ', 'DET', 'ADJ', 'NOUN', 'SYM', 'ADV', 'DET', 'ADJ', 'ADP', 'DET', 'NOUN', 'PRON', 'PRON', 'AUX', 'SYM', 'ADV', 'ADP', 'DET', 'NOUN', 'AUX', 'PRON', 'ADJ', 'SYM', 'CCONJ', 'SCONJ', 'PRON', 'NOUN', 'AUX', 'SYM', 'VERB', 'DET', 'NOUN', 'ADV', 'ADV', 'ADV', 'PRON', 'DET', 'NOUN', 'ADP', 'PRON', 'NOUN', 'SYM', 'SYM', 'SYM']
lemmas = ['nieuw', 'concept', 'in', 'gent', ',', 'maar', 'dat', 'kunnen', 'volgens', 'mij', 'toch', 'goed', '.', 'de', 'veel', 'cornflakes', 'zijn', 'gewoon', 'de', 'basic', 'soort', '.', 'ook', 'wat', 'duur', 'voor', 'de', 'hoeveelheid', 'die', 'je', 'krijgen', ',', 'vooral', 'met', 'de', 'topping', 'zijn', 'ze', 'zuinig', '.', 'en', 'als', 'je', 'ontbijt', 'aanbieden', ',', 'geven', 'de', 'mens', 'dan', 'toch', 'ook', 'wat', 'meer', 'keuze', 'voor', 'hun', 'koffie', '.', '.', '.']

Het opstellen van deze lijsten kost wel wat tijd.  

<div style='color: #690027;' markdown="1">
    <h2>5. Sentiment lexicon matching</h2> 
</div>

Nu je review *gepreprocessed* is, kan je het **sentiment bepalen** met behulp van het sentiment lexicon dat je ter beschikking hebt.<br>
Ook hier werk je in twee grote stappen: je bekijkt het sentiment lexicon en je past daarna matching toe op de review.

Je zoekt de lemma's op in het sentimentlexicon.

In [24]:
# staat "nieuw" in het lexicon?
"nieuw" in lexicondict

True

Het staat er dus in. Vraag de part-of-speech tag ("postag") en de polariteit ("polarity") van "nieuw" op.

In [25]:
lexicondict["nieuw"]["postag"] 

['ADJ']

In [26]:
lexicondict["nieuw"]["polarity"] 

[0.575]

Voor de sentimentanalyse maak je nu een lijst van de **sentimentwoorden**: de tokens die in het lexicon voorkomen (`lexiconmatches`). Je maakt ook een lijst met de polariteiten van deze tokens (`polariteiten`). 

Je overloopt daarvoor een voor een alle lemma's. Voor elk lemma dat in het lexicon staat, controleer je de woordsoort, part-of-speech tag. Bij een correcte woordsoort voeg je het overeenkomstige token toe aan de lijst met tokens en de overeenkomstige polariteit aan de lijst met polariteiten. 

Voor 'nieuw' betekent dit dat 'nieuw' wordt toegevoegd aan de lijst `lexiconmatches` en '0.575' aan de lijst `polariteiten`.

Tot slot tel je alle polariteiten op bij elkaar. De som, `sum(polariteiten)`, geeft het sentiment van de review. 

In [27]:
# zoek lexicon matches in de review
lexiconmatches = []       # lijst tokens gevonden in lexicon
polariteiten = []         # lijst polariteiten van gevonden tokens  

i = 0      # index;  index = 0 komt overeen met eerste lemma em eerste postag
for lemma in lemmas:
    if lemma in lexicondict.keys():  # sleutels zijn woorden aanwezig in lexicon
        if postags[i] in lexicondict[lemma]["postag"]: # alleen wanneer het lemma en de POS-tag overeenkomen, is er een match (zie bv. 'fout als ADJ en 'fout' als NOUN)
            lexiconmatches.append(tokens[i])           # overeenkomstig token toevoegen aan lijst lexiconmatches
            polariteiten.append(sum(lexicondict[lemma]["polarity"]))   # overeenkomstige polariteit toevoegen aan lijst polariteiten
    i = i + 1  # ga over naar volgende lemma, dus lemma met als index eentje meer          

# toon eindbeslissing voor deze review: de som van alle polariteiten
if sum(polariteiten) > 0:
    sentiment = "positief"
elif sum(polariteiten) == 0:
    sentiment = "neutraal"
elif sum(polariteiten) < 0:
    sentiment = "negatief"
print("Het sentiment van de review is: " + sentiment)    

Het sentiment van de review is: positief


<div style='color: #690027;' markdown="1">
    <h2>6. Sentiment lexicon matching: achter de schermen</h2> 
</div>

Bekijk even de sentimentwoorden en hun polariteit door ze op te vragen.

In [28]:
print(lexiconmatches)
print(polariteiten)

['nieuw', 'kan', 'beter', 'gewoon', 'duur', 'voor', 'zuinig', 'geef', 'mensen', 'voor']
[0.575, 1.0, 1.775, 0.5, -3.0666666666666664, 1.0, 0.0, 1.0, 0.0, 1.0]


Je kan in de gegeven review de **sentimentwoorden highlighten**: groen voor een positieve polariteit, rood voor een negatieve en blauw voor een neutrale. 

Je vertrekt daarvoor van de oorspronkelijke tekst. In deze tekst vervang je de tokens die sentimentwoorden waren door zichzelf op een gekleurde achtergrond. De niet-sentimentwoorden laat je ongemoeid.

In [29]:
# lexicon matches in de review highlighten
review_highlighted = review_spatie    # neem review waarin spaties werden aangebracht
         
i = 0   # index
for token in lexiconmatches:
    if polariteiten[i] > 0: # overeenkomstige polariteit is positief
        review_highlighted = review_highlighted.replace(" " + token + " ", " " + Back.GREEN + token + Back.RESET + " ")   # positieve token groen markeren 
    elif polariteiten[i] == 0.0: # overeenkomstige polariteit is neutraal
        review_highlighted = review_highlighted.replace(" " + token + " ", " " + Back.BLUE + token + Back.RESET + " ")    # neutrale token blauw markeren
    elif polariteiten[i] < 0: # negatieve polariteit
        review_highlighted = review_highlighted.replace(" " + token + " ", " " + Back.RED + token + Back.RESET + " ")     # negatieve token rood markeren
    i = i + 1

print(review_highlighted)

 nieuw concept in gent ,  maar dat kan volgens mij toch beter .  de meeste cornflakes waren gewoon de basic soorten .  ook wat duur voor de hoeveelheid die je krijgt ,  vooral met de toppings zijn ze zuinig .  en als je ontbijt aanbiedt ,  geef de mensen dan toch ook wat meer keuze voor hun koffie . 


<div style='color: #690027;' markdown="1">
    <h2>7. Oefening</h2> 
</div>

Bekijk de volgende reviews. 

-  Kies er een uit.
-  Welk sentiment koppel jij eraan?
-  Leidt het lexicon tot hetzelfde resultaat? 

> Sunrisewater is net wat België nodig heeft. Met het obesitasprobleem dat toch wel aan een opmars bezig is, kunnen we alle initiatieven gebruiken om de jeugd weer gewoon water te laten drinken in plaats van die Amerikaanse bucht! Het smaakt geweldig en wat nog beter is, is dat je het gewoon op elke straathoek kan vinden! Echt geweldig! Vooral de pink and yellow is ten zeerste aan te raden.

>  Salé & Sucré staat bekend voor zijn super lekkere en originele cocktails, helaas was er geen alcoholvrije variant te verkrijgen. Onze BOB van dienst moest het dan maar bij frisdrank houden.

>  Het was superleuk om eens te mogen proeven van de Filipijnse keuken. De gerechten zaten goed in elkaar, de porties waren zeker groot genoeg en de smaken zaten helemaal goed. Voor herhaling vatbaar!

>  Gezellige sfeer, lekkere koffie en een mooi interieur. De combinatie van een studiebar en een babbelbar is een geniaal idee! Studeren met een lekker bakkie koffie, een overheerlijk hapje en samen met andere studenten, werkt enorm motiverend. Het interieur is enorm rustgevend met weinig afleiding, waardoor ik nog nooit zoveel heb kunnen doen!

> Wow, wat een coole restaurants! En het eten is er megalekker.

<div class="alert alert-box alert-success">
Proficiat, je hebt geleerd hoe een regelgebaseerd systeem voor sentimentanalyse werkt!
</div>

<img src="../.images/ChatbotSentiment/cclic.png" alt="Banner" align="left" style="width:100px;"/><br><br>
Notebook Chatbot, zie <a href="http://www.aiopschool.be">AI Op School</a>, van C. Van Hee, V. Hoste, F. wyffels, Z. Van de Staey & N. Gesquière is in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>. 